### ISSUE TITLE
Latest build?

### ISSUE THREAD

**Issue by b2cbd** [EionRobb/purple-gowhatsapp#3]
Where to download DLL file?

**Comment by EionRobb** [comment]
https://github.com/hoehermann/purple-gowhatsapp/issues/1#issuecomment-474168496

It's not built from the very latest 'master' though, but I'll see if there's a way to automate it - that dll does work though

**Comment by b2cbd** [comment]
@EionRobb  I tried to add an account using the 13mb dll. I used this username format: 49123456789@s.whatsapp.net.

Showing this error:
> error during login: error decoding login resp: json: cannot unmarshal array into Go value of type map[string]interface {}

**Comment by EionRobb** [comment]
I can't really support the plugin itself, I'm just attempting to get it to compile on Windows.

That said, it looks like you're supposed to just use the phone number, not the old @s format, as that was used by the old plugin, so your setup should just have 49123456789 for the username and "WhatsApp (HTTP)" for the protocol, eg 
![Annotation 2019-11-15 224617](https://user-images.githubusercontent.com/1063865/68933633-f19beb80-07f9-11ea-9b0c-5b0f1fdb24f4.png)

**Comment by EionRobb** [comment]
I've attempted to make a build from the latest master, it compiles but I don't know if it works
[libgowhatsapp.zip](https://github.com/EionRobb/purple-gowhatsapp/files/3850837/libgowhatsapp.zip)

**Comment by b2cbd** [comment]
@EionRobb  Thanks for the new build, unfortunately, it doesn't work! I'm using Pidgin 2.13.0, official latest build on Windows 10 x64. All other protocols including facebook, skype, google works fine!

I tried old build with or without **@s format**, result is same.

> error during login: error decoding login resp: json: cannot unmarshal array into Go value of type map[string]interface {}

And also tried recent build provided by you,  it shows Online... nothing happens after that. No window popups with QR code. In pidgin debug window, it shows this:

> (16:13:03) account: Connecting to account 88019XXXXXXX.
> (16:13:03) connection: Connecting. gc = 37A8XXXX

Any wild guess, what's wrong here?

**Comment by hoehermann** [comment]
Regarding username: WhatsApp Web does not actually care about the user's phone number. As a result, the plug-in does not need it. You can input anything for the username. Though I suspect the empty string could lead to problems within Pidgin.

-----

    error during login: error decoding login resp: json: cannot unmarshal array into Go value of type map[string]interface {}

This error occurred if an old version of Rhymen's go-whatsapp was present during compilation. I did not observe it with builds against recent versions.

-----

> it shows Online... nothing happens after that. No window popups with QR code.

I just tested. It took quite long (about half a minute) until the QR code popped up, but it worked. I am sorry, I cannot reproduce this problem on my system. :|

**Comment by b2cbd** [comment]
@hoehermann  I waited for more than an hour, nothing happens! I've disabled all the plugins and other accounts, in case of any conflict there, still no luck!

So I put a random number as username and create a new WhatsApp HTTP account. It says connecting... Nothing happens after that.

In pidgin debug window I see only these relevant lines:
> (21:16:07) account: Connecting to account 123456.
> (21:16:07) connection: Connecting. gc = 38EXXXX

In the purple directory, I can see a newly created directory - `gowhatsapp` and a `login` file inside which contains hex code.

Is there any other way to debug? or manually create an account by editing files?

**Comment by hoehermann** [comment]
The file with the name `login` actually contains a PNG image. I shows the QR code. Delete the file, then reconnect. As soon as the file appears in the directory, open it. Any browser should be able to display it. IrfanView should do so, too. Scan the QR code with your phone and the log-in procedure should complete. I have no idea why the QR code is not shown to you via conversation window.

The plug-in wrongfully shows as "connected", not as "connecting". I have not been able to figure out what causes this bug, yet. I created the issue https://github.com/hoehermann/purple-gowhatsapp/issues/24 so I don't forget about it.

Apart from that, there is no easy way to have debug output. I commented out most debug output rather than making it configurable. I should probably do so.

**Comment by b2cbd** [comment]
@hoehermann Thanks for adding the issue. So I tried in a manual way.

i. Opened `Pidgin` and waited for other accounts to logged in
ii. Created a new `WhatsApp HTTP` account from account manager
iii. It created a `login` file
iv. Ran a script to open the `QR code`

> echo f | xcopy /f /y login login_copy
> Ren "login_copy" "login.png"
> start login.png

v. Scanned with `Whatsapp` and it successfully added a new entry `rhymen/go-whatsapp`, all this took only 6 seconds
vi. After that nothing happens!  `Available` in the status changer button and `Connecting Icon` in the system-tray

![Capture](https://user-images.githubusercontent.com/6870050/69001940-ae34a080-0911-11ea-9fcc-64d446307dc1.PNG)




vii. I waited for 15 minutes, nothing happened - no incoming message came or buddy list added
viii. I restarted Pidgin, it modified the `login` file, so I guess manual login is pointless from here.
ix. `Available - Connecting` in the status changer button and `Connecting Icon` in the system-tray
x. Waited another 15 minutes, no error thrown.

Whole-time I only found these relevant lines in `Pidgin Debug Window`

> (08:02:46) account: Connecting to account 123456789.
> (08:02:46) connection: Connecting. gc = 35XXXXX

and if it helps, in network-packet-analyzer - I can see `Pidgin` made a bunch of successful requests on `mmx-ds.cdn.whatsapp.net`

**Comment by EionRobb** [comment]
My guess is the `pipe()` change I used to make the plugin compile on Windows doesn't actually work.  I might have to revert to the commit before https://github.com/hoehermann/purple-gowhatsapp/commit/d64c9c9ade21a711e5e68703a8d26cc381867f24 as a workaround?

**Comment by EionRobb** [comment]
Thanks to some clever trickery from @hoehermann I've built a dll that doesn't use any `pipe()`s but instead gets the Go library to call directly into C-land, if someone is able to test it:

[libgowhatsapp.zip](https://github.com/EionRobb/purple-gowhatsapp/files/3861418/libgowhatsapp.zip)

**Comment by rojmiller** [comment]
I have been following the Windows versions of this for quite a while, and I am happy to report that this dll works!!! I now have WhatsApp in Pidgin in Windows 10!

A problem, however - maybe nothing to do with the fork though. What I get popping up in Windows is not the QR code, but a link to the file which contains the code (i.e., login). I then have 20 seconds to scan the code! A real pain as the file has no extension (and I don't want to set all files without extensions to open in IfranView). After several tries I finally got a successful scan, and was logged on to WhatsApp in Pidgin. FYI I just used my phone number as my username.

**Comment by b2cbd** [comment]
@rojmiller You can use the bat script I provided above, it will open the `QR Code` in windows for scanning with just 1-Click with default image viewer. And, you can add anything as username, it doesn't matter here.

@EionRobb Thanks for all the hard work guys! Okay, I've installed the new DLL. `QR code` system message pops up fine. I can add an account now. After adding an account hundreds of chat windows opened.  I guess this is for the first time only to create a history log. For texts, it worked. For images, it creates some below issues... and every time I restart pidgin all those windows will open where an image is needed to download. I tried to disable `Download Media` from account settings, then it opens one chat window mentioning `[File download disabled in settings.]` many-many times.

```
A media message (ID I/fb36) was received, but ID looks not sane – downloading skipped.

A media message (ID AE13F59253CE1D29) was received, but the download failed: Get https://mmi253.whatsapp.net/d/c3Xd-yo/AvBb0T9NFv7.enc: dial tcp: lookup mmi253.whatsapp.net: no such host

A media message (ID 41417191CB3F) was received, but the download failed: download failed with status code 400

A media message (ID EF1E6FCF806FD) was received, but the download failed: download failed with status code 404
```
Update: These errors show only deleted photos from gallery/file but still listed in chat thread.

**and most important issues here,  Whatsapp contacts are not showing in the `Buddy List`. I tried to enable `display all contacts as online / show offline buddies / show empty groups`, no luck! I can initiate a chat from `Buddies > New Instant Messages` menu, contact phone number autocomplete works there, though it's impossible to remember all of my contacts phone number. Chat window also shows number@s.whatsapp.net.**

**Comment by rojmiller** [comment]
@b2cbd Thank you. Good old Windows batch files. Takes me WAY back...

I am getting none of the chat windows opening, even in an account where I know there are a lot of photos. In fact, no previous logs are available, just the logs from any chat since the dll was installed. Weirdly, after a conversation is initiated I get one log of the last chat (pre-new dll) from that contact. So maybe a general Pidgin setting is causing those log windows to open?

I also have to manually add contacts to the Buddy list, because none show up for me either.

@b2cbd  You will have to add an alias to each contact to get the chat window to show real names...

**Comment by b2cbd** [comment]
@rojmiller How are you adding contacts manually?

**Comment by rojmiller** [comment]
"Buddies"-->"Add Buddy", choose your WhatsApp account, enter the phone number for the buddy, enter their real name as the Alias, and add them to the appropriate group.

**Comment by b2cbd** [comment]
@rojmiller  Thanks! Though this will take ages for me, I've like 2000 contacts.

So I made it a bit easier for my purpose. Firstly I exported WhatsApp contact from android using this [app](https://play.google.com/store/apps/details?id=vsv.komishsoft.exportcontactsforwhatsapp&hl=en).  Then I used this dirty-code to add new buddies in pidgin.

```
package main

import (
	"bufio"
	"encoding/csv"
	"io"
	"io/ioutil"
	"log"
	"os"
	"strings"
)

func WriteToFile(filename string, data string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	_, err = io.WriteString(file, data)
	if err != nil {
		return err
	}
	return file.Sync()
}

func main() {
	csvFile, _ := os.Open("whatsapp.csv")
	reader := csv.NewReader(bufio.NewReader(csvFile))
	replacer := strings.NewReplacer("+", "", " ", "", "-", "")

	contents := "		<group name='Whatsapp'>\n"
	contents += "			<setting name='collapsed' type='bool'>1</setting>\n"

	for {
		line, err := reader.Read()
		if err == io.EOF {
			break
		} else if err != nil {
			log.Fatal(err)
		}

		contents += "			<contact>\n"
		contents += "				<buddy account='123456789' proto='prpl-hehoe-gowhatsapp'>\n"
		contents += "					<name>" + replacer.Replace(line[1]) + "@s.whatsapp.net</name>\n"
		contents += "					<alias>" + line[0] + "</alias>\n"
		contents += "				</buddy>\n"
		contents += "			</contact>\n"
	}
	contents += "		</group>\n"
	contents += "	</blist>"

	pidgin, _ := ioutil.ReadFile("blist.xml")
	output := strings.Replace(string(pidgin), "	</blist>", contents, -1)
        WriteToFile("blist.xml.new", output)
}
```
Finally, replaced the `.purple/blist.xml` file with `blist.xml.new`. also needed to replace `123456789` with provided username/number.

@hoehermann For incoming photos, it shows the file local URL, other protocols like purple-facebook can show the image in the chat window. This could be a great enhancement.

**Comment by hoehermann** [comment]
@EionRobb Thank you for your efforts. You are awesome! 🎉

@b2cbd and @rojmiller:  
Thank you for your comments. Your input is appreciated. Based on it, I have put these suggestions to my list of requested enhancements:

* If file download is disabled, mention file only once.  
  In the meantime, a workaround is described in the [wiki](https://github.com/hoehermann/purple-gowhatsapp/wiki/Common-Problems).

This already is on the list:

* Reintroduce inline image display.
  The code is actually already there, but file downloads handling is bit wonky in general. I could not decide whether to inline images or not. I should probably make it user configurable.

This is a known limitation – I need [go-whatsapp](https://github.com/Rhymen/go-whatsapp) to overcome it:

* Downloading (some also say "syncing") contacts from phone is an enhancement already requested at https://github.com/hoehermann/purple-gowhatsapp/issues/12.   
  Good you found out how to add contacts to the buddy list manually and have their names as aliases.

That should be all.

**Comment by rojmiller** [comment]
@b2cbd I just noticed on incoming messages the contacts info is not just the phone number - it is (phone number)@s.whatsapp.net. It may have to be in that form to avoid duplication, because thats the way incoming messages are formatted.

**Comment by b2cbd** [comment]
@hoehermann  Thanks for adding those in the queue! I want to add few more issues.

i. Incoming `Sticker` is not showing, not even text. `Purple-facebook` can show stickers, it will be a great enhancement too! I can see `go-whatsapp` have merged sticker download - [d7754af](https://github.com/Rhymen/go-whatsapp/commit/d7754af4a6b7209d88132b5e498c98f12fb67f70)
ii After writing a message in the chat window, when I press enter to send - it shows a `delay/latency`(half to one second) to complete. I don't see this delay in any other protocols. I'm not sure it's windows specific or for all platforms.
iii. I see this protocol has a built-in `auto-reply` option in account settings. I tried to enable/disable it - but it doesn't work.
iv. I use an old but rock-solid plugin `Autoreply - Sadrul Habib Chowdhury`, it works flawlessly for any other protocols including purple-facebook. But it doesn't work with Whatsapp! I'm not sure what's special for Whatsapp here. Basically, it throws an autoreply message to every incoming message when I select `Away` as my global status. Are there any issues around `Away vs. Available` for Whatsapp protocol?


@rojmiller Yes, it has to be `number with country code( no + ) @ s.whatsapp.net` Already in my dirty [code](https://github.com/EionRobb/purple-gowhatsapp/issues/3#issuecomment-555356491).

`contents += "<name>" + replacer.Replace(line[1]) + "@s.whatsapp.net</name>\n"
`

**Comment by hoehermann** [comment]
@rojmiller You are correct. This is what WhatsApp uses as an ID. I am using them as they are provided by WhatsApp.

@b2cbd Thank you for your suggestions.

1. In case stickers behave anything like images, I want the plug-in to support them.  
  Please also be aware that Stickers are so new, go-whatsapp did not even support them a fortnight ago. Please do not expect quick implementation of new features (keep in mind that I am doing this for fun).
1. This already is on the list (see the readme), but with a very low priority.  
  The reason behind this is that I wait for the send request to complete. Other plug-ins seem to do it the very same way. I assume, their protocols or back-ends are simply not as sluggish as WhatsApp's. Sending messages asynchronously is relatively hard to implement correctly, while not gaining much usability. Mind that you can still start to type your next message – you just cannot see the text instantly.
1. You are mistaken. The plug-in automatically sends so-called "receipts", marking an incoming message as "received and displayed", making the check-mark turn blue on the sender's phone. In case I misunderstand you, please clarify what option you are talking about.
1. I do not know about the `Autoreply` plug-in and never ran tests against it. [Apparently](https://github.com/Javran/purple-plugin-pack/blob/master/autoreply/autoreply.c#L156), it works only on accounts which are set to "Away". purple-gowhatsapp however only "Online" and "Offline" (disconnected). I add this to the list of requested enhancements.

**Comment by b2cbd** [comment]
@hoehermann 

I was talking about this `autoreply` on [#3 and #4](https://github.com/EionRobb/purple-gowhatsapp/issues/3#issuecomment-555814663) 

![Capture](https://user-images.githubusercontent.com/6870050/69300602-e6075500-0c3d-11ea-9982-14cb63e42d6a.PNG)

Updated: Anyways, as you said it conflicts with only online/offline status. It will work when `Away` status will be implemented. We really appreciate all the good work you've done here. Thank you!

**Comment by hoehermann** [comment]
@b2cbd The options high-lighted in the screenshot are injected into purple-gowhatsapp's options by the auto-reply plug-in. I have no control over them. However, there apparently is a [modified auto-reply](https://sourceforge.net/projects/autoreplymodif/) plug-in (not tested nor hereby advertised by me) which mentions a support for more statuses.

**Comment by b2cbd** [comment]
@hoehermann  Yap, it works. Thank you! :)

**Comment by b2cbd** [comment]
> @b2cbd The options high-lighted in the screenshot are injected into purple-gowhatsapp's options by the auto-reply plug-in. I have no control over them. However, there apparently is a [modified auto-reply](https://sourceforge.net/projects/autoreplymodif/) plug-in (not tested nor hereby advertised by me) which mentions a support for more statuses.

@hoehermann I thought this solved the issue, actually not. It doesn't work for `Away`, nor `Available`. It sends an auto-reply message on this scenario sometimes, not always:

```
i.   Whatsapp logged in Pidgin
ii.  Received message in both Pidgin and Phone
iii. Open Whatsapp in Phone and Replied
iv.  It sent an autoreply immediately(for my own message)
```
So this is happening for something else, not for the status! I can confirm this plugin(modified or original) perfectly works for Facebook/Google/Skype/Others. I guess  WhatsApp Protocol breaks the standard somewhere.

**Comment by b2cbd** [comment]
@EionRobb Can you please re-build latest [commit](https://github.com/hoehermann/purple-gowhatsapp/commit/e0adb32fcec1571de62b5fbf2dec9a7034d4bf74)? Thanks in advance!

**Comment by EionRobb** [comment]
@b2cbd the build at https://github.com/hoehermann/purple-gowhatsapp/wiki/Windows-Build should be the latest

**Comment by rojmiller** [comment]
I am using the latest build, and have noticed a small problem - normally, the messages come in fine, in the order they were sent. But if my computer is off or asleep when messages are sent, when Pidgin starts up the messages come in and are shown in reverse order! (and appear in that order in the logs)

**Comment by hoehermann** [comment]
@rojmiller This is a known problem (bullet point "Sort old messages by date" in [the readme](https://github.com/hoehermann/purple-gowhatsapp/blob/master/README.md)). It annoys me, too, every time I log on. And then I think about what a hassle it is to collect an unknown number of messages, delaying all of them for an uncertain amount of time (which depends on internet service quality, possibility of attachment presence and download speed) until they can be sorted. I might change this one day, but probably not today. It is not a big amount of work, but (given the current state of go-whatsapp) it cannot be reliable. At least the time-stamps are correct so you can infer the message order by looking at them.
